exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/entities/CompositeIdTableEntityTest.kt

```diff
@@ -0,0 +1,556 @@
+import org.jetbrains.exposed.sql.tests.currentTestDB
+import org.jetbrains.exposed.sql.tests.shared.assertEquals
+import org.jetbrains.exposed.sql.tests.shared.assertTrue
+import org.jetbrains.exposed.sql.tests.shared.expectException
+import org.junit.Test
+import java.util.*
+import kotlin.test.assertIs
+import kotlin.test.assertNotNull
+
+// SQLite excluded from most tests as it only allows auto-increment on single column PKs.
+// SQL Server is sometimes excluded because it doesn't allow inserting explicit values for identity columns.
+@Suppress("UnusedPrivateProperty")
+class CompositeIdTableEntityTest : DatabaseTestsBase() {
+    // CompositeIdTable with 2 key columns - int & uuid (both db-generated)
+    object Publishers : CompositeIdTable("publishers") {
+        val pubId = integer("pub_id").autoIncrement().entityId()
+        val isbn = uuid("isbn_code").autoGenerate().entityId()
+        val name = varchar("publisher_name", 32)
+
+        override val primaryKey = PrimaryKey(pubId, isbn)
+    }
+
+    class Publisher(id: EntityID<CompositeID>) : CompositeEntity(id) {
+        companion object : CompositeEntityClass<Publisher>(Publishers)
+
+        var name by Publishers.name
+    }
+
+    // IntIdTable with 1 key columns - int (db-generated)
+    object Authors : IntIdTable("authors") {
+        val publisherId = integer("publisher_id")
+        val publisherIsbn = uuid("publisher_isbn")
+        val penName = varchar("pen_name", 32)
+
+        // FK constraint with multiple columns is created as a table-level constraint
+        init {
+            foreignKey(publisherId, publisherIsbn, target = Publishers.primaryKey)
+        }
+    }
+
+    class Author(id: EntityID<Int>) : IntEntity(id) {
+        companion object : IntEntityClass<Author>(Authors)
+
+        var publisher by Publisher referencedOn Authors
+        var penName by Authors.penName
+    }
+
+    // CompositeIdTable with 1 key column - int (db-generated)
+    object Books : CompositeIdTable("books") {
+        val bookId = integer("book_id").autoIncrement().entityId()
+        val title = varchar("title", 32)
+        val author = optReference("author_id", Authors)
+
+        override val primaryKey = PrimaryKey(bookId)
+    }
+
+    class Book(id: EntityID<CompositeID>) : CompositeEntity(id) {
+        companion object : CompositeEntityClass<Book>(Books)
+
+        var title by Books.title
+        var author by Author optionalReferencedOn Books.author
+        val review by Review backReferencedOn Reviews.book
+    }
+
+    // CompositeIdTable with 2 key columns - string & long (neither db-generated)
+    object Reviews : CompositeIdTable("reviews") {
+        val content = varchar("code", 8).entityId()
+        val rank = long("rank").entityId()
+
+        // FK constraint with single column can be created as a column-level constraint
+        val book = reference("book_id", Books.bookId)
+
+        override val primaryKey = PrimaryKey(content, rank)
+    }
```

Comment:
A point to note:

Using the following tables are synonymous:
```kt
object Tester1 : IntIdTable() {
    // more columns
}

object Teste2: IdTable<Int>() {
    override val id = integer(columnName).autoIncrement().entityId()
    // more columns
    override val primaryKey = PrimaryKey(id)
}
```

Trying to do something similar will not work to replace `CompositeIdTable`, mostly because of how the `id` column is defined but also because of internal checks. Should included this behavior be a must?
```kt
object Tester3 : IdTable<CompositeID>() {
    override val id = //...
    // pk columns
    override val primaryKey = PrimaryKey(...)
}
```